### ISSUE TITLE
dependent-map broken on 8.2

### ISSUE THREAD

**Issue by luite** [ghcjs/ghcjs#646]
the following code prints
```
[LeftTag :=> "def"]
```
instead of the expected
```
[LeftTag :=> "abc",RightTag :=> "def"]
```

It looks like this is the reason that reflex doesn't properly propagate events.

```haskell
{-# LANGUAGE GADTs, MultiParamTypeClasses #-}
{-# OPTIONS_GHC -O0 #-}
import qualified Data.Dependent.Map as DMap
import Data.Dependent.Map (DMap, DSum (..), GCompare (..))
import Data.Dependent.Sum
import Data.GADT.Show
import Data.GADT.Compare
import Data.Functor.Identity

-- | Tag type for 'Either' to use it as a 'DSum'.
data EitherTag l r a where
  LeftTag :: EitherTag l r l
  RightTag :: EitherTag l r r

instance GEq (EitherTag l r) where
  geq a b = case (a, b) of
    (LeftTag, LeftTag) -> Just Refl
    (RightTag, RightTag) -> Just Refl
    _ -> Nothing

instance GCompare (EitherTag l r) where
  gcompare a b = case (a, b) of
    (LeftTag, LeftTag) -> GEQ
    (LeftTag, RightTag) -> GLT
    (RightTag, LeftTag) -> GGT
    (RightTag, RightTag) -> GEQ

instance GShow (EitherTag l r) where
  gshowsPrec _ a = case a of
    LeftTag -> showString "LeftTag"
    RightTag -> showString "RightTag"

instance (Show l, Show r) => ShowTag (EitherTag l r) Identity where
  showTaggedPrec t n (Identity a) = case t of
    LeftTag -> showsPrec n a
    RightTag -> showsPrec n a

main :: IO ()
main = do
  let ea = Identity "abc"
      eb = Identity "def"
  print $ DMap.toList (DMap.fromList [LeftTag :=> ea, RightTag :=> eb])
```

**Comment by luite** [comment]
the pattern match on the GADT appears to go wrong

**Comment by luite** [comment]
I think 50d7ed0eea427c74893a1e0b76bc5e3fb8231379 fixes this. Whew, this took me a bit more time to track down than I planned...

Now to merge/push the other fixes I've been working on and hopefully the 8.2 branch is usable.

**Comment by ryantrinkle** [comment]
Awesome! Thanks so much for tracking this down - it would have taken me
forever!
